internal/command/jsonformat/change/list.go

```diff
@@ -0,0 +1,82 @@
+package change
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/hashicorp/terraform/internal/command/format"
+	"github.com/hashicorp/terraform/internal/plans"
+)
+
+func List(elements []Change) Renderer {
+	return &listRenderer{
+		displayContext: true,
+		elements:       elements,
+	}
+}
+
+func NestedList(elements []Change) Renderer {
+	return &listRenderer{
+		elements: elements,
+	}
+}
+
+type listRenderer struct {
+	NoWarningsRenderer
+
+	displayContext bool
+	elements       []Change
+}
+
+func (renderer listRenderer) Render(change Change, indent int, opts RenderOpts) string {
+	if len(renderer.elements) == 0 {
+		return fmt.Sprintf("[]%s%s", change.nullSuffix(opts.overrideNullSuffix), change.forcesReplacement())
+	}
+
+	elementOpts := opts.Clone()
+	elementOpts.overrideNullSuffix = true
+
+	unchangedElementOpts := opts.Clone()
+	unchangedElementOpts.showUnchangedChildren = true
+
+	var unchangedElements []Change
+	renderNext := false
+
+	var buf bytes.Buffer
+	buf.WriteString(fmt.Sprintf("[%s\n", change.forcesReplacement()))
+	for _, element := range renderer.elements {
+		if element.action == plans.NoOp && !renderNext && !opts.showUnchangedChildren {
+			unchangedElements = append(unchangedElements, element)
+			continue
+		}
+		renderNext = false
+
+		if renderer.displayContext {
```

Comment:
I'm finding the logic around displaying contextual unchanged elements a bit difficult to follow, in part because it's procedural, and in part because it's Friday. Maybe some commentary on how this works would help.

It might also be neat to have a more data based approach here: take the list and the options and create a temporary set of data to be rendered, then separately render each line. This is a pie-in-the-sky idea and may not actually help, so just noting it in case it somehow inspires further thought.

Reply:
I've added additional comments trying to explain things better here.

I think if I was going to redo the approach to rendering this, I'd want to make all the other collection types consistent. So I will think about this, and if I think I can improve it then I'll apply the change to all the collections in a single refactoring PR.